core/src/main/java/com/datastax/oss/driver/internal/core/context/NettyOptions.java

```diff
@@ -62,6 +66,45 @@
    */
   ByteBufAllocator allocator();
 
+  /**
+   * Sets the socket options on a newly created client bootstrap. This is invoked each time the
+   * driver opens a new channel.
+   *
+   * <p>Implementation note: this is very similar to {@link #afterBootstrapInitialized}; the only
+   * reason to have two separate methods is backward compatibility.
+   *
+   * @param bootstrap the bootstrap to configure.
+   * @param defaultProfile the default profile of the driver's configuration (where the values to
+   *     set will typically be read from).
+   * @since 4.1.0
+   */
+  default void setSocketOptions(Bootstrap bootstrap, DriverExecutionProfile defaultProfile) {
```

Comment:
One thing sort of awkward about this is that this method provides `DriverExecutionProfile` where the others don't. 

I was thinking maybe we could overload `afterBootstrapInitialized` with another method providing the profile as a parameter, but it gets sort of confusing to have two methods with the same name, and as you explained we don't want to break existing custom implementations which implicitly depended on the options to be set otherwise, so I think what you did here is probably our best option.

Reply:
> One thing sort of awkward about this is that this method provides `DriverExecutionProfile` where the others don't.

Yes, that's the only way I found to provide a default implementation.

Reply:
Do we really expect people extending or implementing `NettyOptions`? This being an internal class, I would be in favor of being more agressive with backwards compatibility.

Reply:
Very good point, for some reason I was assuming it was in the public API and didn't even check the package name.
You're right, it's fair to be a bit more aggressive here, I'll do the changes.